File: courseschedule.js
//Objective is to see if it's possible to take a certain number of classes,
//given each class has a set of prerequisite classes.

let numCourses = 2, prerequisites = [[1,0],[0,1]]


//O(n) solution that does a dfs traversal through each vertex to see if 
//there's a cycle, in which case the course schedule is invalid

let graph = new Map()
let visited = new Set()
let visiting = new Set()

//Make the adjacency list
for (let [v,e] of prerequisites) {
    if (!graph.has(v)) {
        graph.set(v, [e])
    } else {
        let edges = graph.get(v)
        edges.push(e)
        graph.set(v, edges)
    }
}

//Check every vertex for a cycle
for (let [v,e] of graph) {
    if (dfs(v)) {
        return false
    }
}

return true

function dfs(vertex) {
    visiting.add(vertex)
    let edges = graph.get(vertex)
    
    if (edges) {
        for (let edge of edges) {
            //If the edge has been visited already, skip
            if (visited.has(edge)) {
                continue
            }
            
            //If we're visiting the same node in the same loop,
            //there is a cycle
            if (visiting.has(edge)) {
                return true
            }
            
            //If a neighbor of the edge also has a cycle,
            //then the graph is still invalid
            if (dfs(edge)) {
                return true
            }
        }
    }
    
    visiting.delete(vertex)
    visited.add(vertex)
    return false
}

File: courseschedule.java
//Java Solution

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        
        for (int[] edge : prerequisites) {
            if (!graph.containsKey(edge[0])) {
                graph.put(edge[0], new ArrayList<Integer>());
            }
            graph.get(edge[0]).add(edge[1]);
        }
        
        Set<Integer> visited = new HashSet<>();
        Set<Integer> visiting = new HashSet<>();
        
        for (Map.Entry<Integer, List<Integer>> vertex : graph.entrySet()) {
            if (dfs(vertex.getKey(), graph, visited, visiting)) {
                return false;
            }
        }
        
        return true;
    }
    
    public boolean dfs(int vertex, Map<Integer, List<Integer>> graph, Set<Integer> visited, Set<Integer> visiting) {
        visiting.add(vertex);
        List<Integer> edges = graph.get(vertex);
        
        if (edges != null) {
            for (int edge : edges) {
                if (visited.contains(edge)) {
                    continue;
                }
                
                if (visiting.contains(edge)) {
                    return true;
                }
                
                if (dfs(edge, graph, visited, visiting)) {
                    return true;
                }
            }
        }
        
        visited.add(vertex);
        visiting.remove(vertex);
        return false;
    }
}